packages/api/src/routes/metaplex-upload.js

```diff
@@ -0,0 +1,158 @@
+import { HTTPError } from '../errors.js'
+import * as cluster from '../cluster.js'
+import { secrets, database } from '../constants.js'
+import { JSONResponse } from '../utils/json-response.js'
+import { toNFTResponse } from '../utils/db-transforms.js'
+import { parseCid } from '../utils/utils.js'
+import {
+  parseJWTHeader,
+  parseJWT,
+  Base64URL,
+  utf8ToUint8Array,
+} from '../utils/jwt.js'
+import { keyFromDID, verifyEd25519Signature } from '../utils/ed25519.js'
+import { DBClient, DBError } from '../utils/db-client.js'
+
+/**
+ * When >2.5MB, use local add, because waiting for blocks to be sent to
+ * other cluster nodes can take a long time. Replication to other nodes
+ * will be done async by bitswap instead.
+ */
+const LOCAL_ADD_THRESHOLD = 1024 * 1024 * 2.5
+
+// TODO: remove when https://github.com/nftstorage/nft.storage/pull/638 merged
+const db = new DBClient(database.url, secrets.database)
+
+/**
+ * Temporary CAR only upload endpoint, authenticated with Metaplex JWT.
+ * @type {import('../bindings').Handler}
+ */
+export async function metaplexUpload(event, ctx) {
+  const { headers } = event.request
+  const authHeader = headers.get('x-web3auth') || ''
+
+  const match = authHeader.match(/^Metaplex (.*)$/)
+  if (!match) {
+    throw new HTTPError('invalid authorization header: ' + authHeader, 401)
+  }
+
+  const token = match[1]
+  const valid = await verifyMetaplexJWT(token)
+  if (!valid) {
+    throw new HTTPError('invalid authorization header', 401)
+  }
+
+  // now that we know the sig is valid, use the JWT payload
+  // as metadata to attach to the upload entry in the db
+  /** @type Record<string, any> */
+  const { iss, req } = parseJWT(token)
+  if (!iss) {
+    throw new HTTPError(
+      'invalid token - required field "iss" not present in payload',
+      401
+    )
+  }
+  if (!req) {
+    throw new HTTPError(
+      'invalid token - required field "req" not present in payload',
+      401
+    )
+  }
+
+  // Get the mapped metaplex user
+  const { user, key } = await validate()
+
+  const blob = await event.request.blob()
+  if (blob.size === 0) {
+    throw new HTTPError('Empty payload', 400)
+  }
+
+  const local = blob.size > LOCAL_ADD_THRESHOLD
+  const { cid, bytes: dagSize } = await cluster.addCar(blob, { local })
+  const { sourceCid, contentCid } = parseCid(cid)
+
+  // confirm that the content cid matches the cid from the signed metadata
+  if (!req.put || typeof req.put.rootCID !== 'string') {
+    throw new HTTPError('invalid token - no root cid in payload', 401)
+  }
+  const signedCid = req.put.rootCID
+  if (signedCid !== sourceCid && signedCid !== contentCid) {
```

Comment:
`contentCid` is the normalised CID. Are you saying that `req.put.rootCID` will _always_ be a v1 base32 encoded CID? There's no way people can put `Qm...` CID's in the CAR?

Reply:
hmm... good point. I should probably use `CID.parse` and `CID.equals` instead.

Reply:
I realized I can just use the same `parseCid` helper and compare `contentCid`s instead of `CID.parse`.